### ISSUE TITLE
Racy/confusing use of KeepaliveRunning

### ISSUE THREAD

**Issue by mvdan** [TykTechnologies/tyk#533]
We have this:

```
var KeepaliveRunning bool

func StartRPCKeepaliveWatcher(engine *RPCStorageHandler) {
        if KeepaliveRunning {
                return
        }
        go func() {
                for {
                        KeepaliveRunning = true
                        RPCKeepAliveCheck(engine)
                        // ...
                }
        }
}
```

which is called only in:

```
StartRPCKeepaliveWatcher(rpcAuthStore)
StartRPCKeepaliveWatcher(rpcOrgStore)
```

This logic seems broken to me.

* Why are we calling the func twice, if the watch loop is supposed to run only once?
* It is racy - the global bool is read in the main goroutine, but set in another goroutine. We might have one or two watch loops.

Found this while looking for global bools that might be racy code like the hot reload logic.

@lonelycode thoughts?

/cc @buger

**Comment by mvdan** [comment]
Perhaps the intent was to have exactly one watch loop per store. In that case, we would need two bools or another mechanism to ensure it properly.

**Comment by mvdan** [comment]
To clarify - the two calls are in `loadApps`, which can be executed any number of times.

**Comment by buger** [comment]
I assume intention was to keep 2 separate keepalive checks per engine, and `KeepaliveRunning` should be variable of `RPCStorageHandler` object.

**Comment by mvdan** [comment]
> 2 separate keepalive checks per engine

I assume you mean one per engine?

> and KeepaliveRunning should be variable of RPCStorageHandler object.

That could also be a way to fix it. My guess is that we used to have only one, so a global bool did the trick (minus the race).

**Comment by buger** [comment]
> I assume you mean one per engine?
yes

**Comment by lonelycode** [comment]
> Perhaps the intent was to have exactly one watch loop per store. In that case, we would need two bools or another mechanism to ensure it properly.

That was the idea, now that it's a singleton it's no longer needed twice

We should internalise it to the store itself, and clean it up - the RPC driver should be handling this for us, but it doesn't do it properly I think, our ELB still kills connections after a period of inactive time
